specifyweb/frontend/js_src/lib/components/FormPlugins/useMoment.tsx

```diff
@@ -0,0 +1,75 @@
+import React from 'react';
+
+import { dayjs, getDateInputValue } from '../../utils/dayJs';
+import { databaseDateFormat } from '../../utils/parser/dateConfig';
+import type { AnySchema } from '../DataModel/helperTypes';
+import type { SpecifyResource } from '../DataModel/legacyTypes';
+import { resourceOn } from '../DataModel/resource';
+
+/**
+ * Get a resource's date field value as a moment.js object
+ */
+export function useMoment(
+  resource: SpecifyResource<AnySchema> | undefined,
+  dateFieldName: string,
+  defaultValue: Date | undefined
+): readonly [
+  ReturnType<typeof dayjs> | undefined,
+  (value: ReturnType<typeof dayjs> | undefined) => void
+] {
+  const syncMoment = React.useCallback(
+    (moment: ReturnType<typeof dayjs> | undefined) => {
+      const value = resource?.get(dateFieldName) ?? undefined;
+      const newMoment =
+        value === undefined
+          ? undefined
+          : dayjs(value, databaseDateFormat, true) ?? undefined;
+
+      // Only change the object instance if it is different
+      return moment === undefined ||
+        newMoment === undefined ||
+        moment.toJSON() !== newMoment.toJSON()
+        ? newMoment
+        : moment;
+    },
+    [resource, dateFieldName]
+  );
+  // Parsed date object
+  const [moment, setMoment] = React.useState<
+    ReturnType<typeof dayjs> | undefined
+  >(() => syncMoment(undefined));
+
+  React.useEffect(() => {
+    if (resource === undefined) return undefined;
+    if (
+      typeof defaultValue === 'object' &&
+      typeof resource === 'object' &&
+      resource.isNew() &&
+      typeof resource?.get(dateFieldName) !== 'string'
+    )
+      resource.set(dateFieldName, getDateInputValue(defaultValue) as never, {
+        silent: true,
+      });
+
+    return resourceOn(
+      resource,
+      `change:${dateFieldName}`,
+      () => setMoment(syncMoment),
+      true
+    );
+  }, [resource, dateFieldName, syncMoment, defaultValue]);
+
+  return [
+    moment,
+    React.useCallback(
+      (moment) => {
+        if (moment?.isValid() !== false)
+          resource?.bulkSet({
+            [dateFieldName]: moment?.format(databaseDateFormat) ?? null,
+          });
+        setMoment(moment);
```

Comment:
On a side note, is there any harm in changing `if (moment?.isValid() !== false)` to  `if (moment?.isValid() === true)`? 
It would be slightly easier to reason about and doesn't change the functionality of the following `setMoment(moment)` which is always being called.

Reply:
> Why was bulkSet used here over set?

great question, and something I should have made clearer in the code

in present form `.set()` is trying to be type safe with field names and argument names
when I was adding that, I was considering the cases when you are dealing with a specific known table, and a hardcoded field name - in those cases this type safety is very useful
but I failed to consider the generic cases like this - the PartialDateUi could be used in any table, and field name is dynamic too, so typing fails

`.bulkSet()` does not have any of this type-safety logic, so that's why I used it here (to avoid a typescript error)
in other places in the code we used an `as never` assertion, which is probably worse than using not-type-checked `.bulkSet()`

we could add a `.genericSet()` that would be exactly like `.set()` but untyped, but since it would be equivalent to `.bulkSet()`, the only benefit is code clarity
but since we are looking to refactor the backbone code, I didn't want to do such changes now in this PR

Reply:
> On a side note, is there any harm in changing if (moment?.isValid() !== false) to if (moment?.isValid() === true)?

Not equivalent.
The current condition is true if moment is valid OR moment is undefined (in which case the field value is set to null)
Your condition is only true if moment is defined AND valid

being explicit like `if(moment === undefined || moment.isValid())` is probably a better idea for readability - thanks for the suggestion, will refactor

Reply:
Ah yes, you are correct my apologies. 
At the moment I was assuming that if moment was undefined, then it is a falsy value and thus the conditions would be equivalent (because `(moment?.isValid() !== false)` would return `false` in that case). 
Somehow slipped my mind we were doing an explicit check (!==) rather than implicitly converting the type (!=) in the comparison :facepalm: